Added diff functionality for Git and SVN

shanept opened the pull request:
Creates a function 'getDiff' for repositories, which returns the raw output of the diff command for the respective repository engine.

sgehrig commented on src/TQ/Git/Repository/Repository.php:
```diff
@@ -903,6 +903,29 @@ public function getStatus()
+     * @param   bool    $staged     Should the diff return for the staged file
+     * @return  string
+     */
+    public function getDiff($file, $staged=false)
```
Sorry for being picky ;-) Could you add some whitespace around the `=`?

```
public function getDiff($file, $staged = false)
```

shanept replied:
Of course!

sgehrig commented:
Generally I'd favor a more structured return value - at least an array of diffs (one per file). But I think it's OK for now as everybody could parse the response to create a structured value.

shanept commented:
I didn't expect people to search for diffs⁣ from multiple files, but now that you mention it I do see the validity of such an approach.

I will add those spaces, for sure.

On 12 Jan 2017, 4:35 PM, at 4:35 PM, Stefan Gehrig <notifications@github.com> wrote:
>Generally I'd favor a more structured return value - at least an array
>of diffs (one per file). But I think it's OK for now as everybody could
>parse the response to create a structured value.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/teqneers/PHP-Stream-Wrapper-for-Git/pull/24#issuecomment-272104748

sgehrig commented:
Perfect! Thanks for your contribution.

shanept commented:
More than happy to help. Thank you

On 12 Jan 2017, 5:04 PM, at 5:04 PM, Stefan Gehrig <notifications@github.com> wrote:
>Perfect! Thanks for your contribution.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/teqneers/PHP-Stream-Wrapper-for-Git/pull/24#issuecomment-272110259

sgehrig commented:
Please don't forget to rebase to current master and squash all commits. THX!

shanept commented:
Ok, I have modified it so you can pass an array of filenames and get an array of diffs for each file, or by not passing any filenames (or null), it will return diffs for all files currently modified.

I have also squashed all commits.

sgehrig commented:
There seems to be a problem with the `TQ\Tests\Svn\Repository\InfoTest::testGetDiff` test when running on PHP 5.3 and 5.4 with PHPUnit 4.8.31 (this is the highest version available for PHP < 7) respectively. The order of the resulting array seems to be different. Rather weird from my perspective.

shanept commented:
Hi Stefan,

That is peculiar. I ran it on PHP7 and a different version of PHPUnit and didn't have a problem, except with some deprecations in older versions. I'll check it out and get back with a fix.
Thanks⁣

Sent from BlueMail ​

On 13 Jan 2017, 4:26 PM, at 4:26 PM, Stefan Gehrig <notifications@github.com> wrote:
>There seems to be a problem with the
>`TQ\Tests\Svn\Repository\InfoTest::testGetDiff` test when running on
>PHP 5.3 and 5.4 with PHPUnit 4.8.31 (this is the highest version
>available for PHP < 7) respectively. The order of the resulting array
>seems to be different. Rather weird from my perspective.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/teqneers/PHP-Stream-Wrapper-for-Git/pull/24#issuecomment-272387495

sgehrig commented:
Strange - indeed. Perhaps it's possible to adjust the test so that the order doesn't matter. 
Which component defines the order in these arrays? Is it the underlying SVN cli command? Or is it something in PHP?

shanept commented:
Stefan, the order is defined by the order in which the⁣ the files are passed to the diff function. Otherwise it is defined by the order in which the getStatus call returns them. Simple fix would be to sort the array by key before comparing them

Sent from BlueMail ​

On 13 Jan 2017, 6:10 PM, at 6:10 PM, Stefan Gehrig <notifications@github.com> wrote:
>Strange - indeed. Perhaps it's possible to adjust the test so that the
>order doesn't matter. 
>Which component defines the order in these arrays? Is it the underlying
>SVN cli command? Or is it something in PHP?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/teqneers/PHP-Stream-Wrapper-for-Git/pull/24#issuecomment-272407468

sgehrig commented:
It indeed seems as if the `getStatus()` method for SVN does not guarantee a strict order - for whatever reason. So sorting might be the best option IMHO.

shanept commented:
Stefan, I have added sorting to the SVN diff.

sgehrig commented:
Looks good. Thanks!

shanept commented:
Awesome, thank you!
